compiler/rustc_trait_selection/src/traits/vtable.rs

```diff
@@ -383,17 +385,37 @@ pub(crate) fn supertrait_vtable_slot<'tcx>(
     let ty::Dynamic(target, _, _) = *target.kind() else {
         bug!();
     };
-    let target_principal = tcx
-        .normalize_erasing_regions(ty::ParamEnv::reveal_all(), target.principal()?)
-        .with_self_ty(tcx, tcx.types.trait_object_dummy_self);
+    let target_principal = target.principal()?.with_self_ty(tcx, tcx.types.trait_object_dummy_self);
 
     // Given that we have a target principal, it is a bug for there not to be a source principal.
     let ty::Dynamic(source, _, _) = *source.kind() else {
         bug!();
     };
-    let source_principal = tcx
-        .normalize_erasing_regions(ty::ParamEnv::reveal_all(), source.principal().unwrap())
-        .with_self_ty(tcx, tcx.types.trait_object_dummy_self);
+    let source_principal =
+        source.principal().unwrap().with_self_ty(tcx, tcx.types.trait_object_dummy_self);
+
+    let infcx = tcx.infer_ctxt().build();
+    let param_env = ty::ParamEnv::reveal_all();
+    let trait_refs_are_compatible =
+        |source: ty::PolyTraitRef<'tcx>, target: ty::PolyTraitRef<'tcx>| {
+            infcx.probe(|_| {
```

Comment:
this ends up doing trait solving with a `FreshTy` type inference variable as the self type as we replace the self type with `tcx.types.trait_object_dummy_self`, doesn't it?

This feels quite broken to me :thinking: can we modify this code to use `ExistentialTraitRef` as much as possible?

Reply:
No, there is no trait solving that will ever be done on a fresh ty, unless we start emitting AliasRelate goals for ?0 = FreshTy.

I can change this routine to relate the ExistentialTraitRefs but that's the only change I'm willing to do. The vtable walking code is very hard to modify to stop using PolyTraitRefs because of the way it operates--I tried.